modules/cudf/src/data_frame.ts

```diff
@@ -253,13 +264,75 @@ export class DataFrame<T extends TypeMap = any> {
    */
   filter(mask: Series<Bool8>) {
     const temp       = new Table({columns: this._accessor.columns});
-    const columns    = temp.gather(mask._col);
+    const columns    = temp.gather(mask._col, false);
     const series_map = {} as SeriesMap<T>;
     this._accessor.names.forEach(
       (name, index) => { series_map[name] = Series.new(columns.getColumnByIndex(index)); });
     return new DataFrame(series_map);
   }
 
+  /**
+   * Join columns with other DataFrame.
+   *
+   * @param props the configuration for the join
+   * @returns the joined DataFrame
+   */
+  join<U extends TypeMap, R extends keyof U>(props: JoinProps<U, R>): DataFrame<any> {
+    const {how, other, on} = props;
+    const nullEquality     = props.nullEquality == undefined ? true : props.nullEquality;
+    const lsuffix          = props.lsuffix == undefined ? '' : props.lsuffix;
+    const rsuffix          = props.rsuffix == undefined ? '' : props.rsuffix;
+    const left             = this.select(on as any).asTable();
+    const right            = other.select(on as any).asTable();
+
+    const joins = {
+      'inner': () => Table.innerJoin(left, right, nullEquality),
+      'outer': () => Table.fullJoin(left, right, nullEquality),
+      'left': ()  => Table.leftJoin(left, right, nullEquality),
+      'right': () => Table.leftJoin(right, left, nullEquality),
+    };
+
+    let [left_gather, right_gather] = joins[how]();
+    if (how == 'right') { [left_gather, right_gather] = [right_gather, left_gather]; }
+
+    let left_result  = this.gather(Series.new(left_gather), true);
+    let right_result = other.gather(Series.new(right_gather), true);
+
+    // for an outer join, we need to to merge the values from both left and right "on" columns
+    if (how == 'outer') {
+      for (const name of on) {
+        const rcol   = right_result.get(name);
+        const lcol   = left_result.get(name as any);
+        const merged = lcol.replaceNulls(rcol as any);
+        left_result  = left_result.assign({[name]: merged});
+      }
+    }
+
+    right_result = right_result.drop(on) as any;
+
+    // if there are no suffices provided, and there are column conflicts, then we
```

Comment:
```suggestion
    // if there are no suffixes provided, and there are column conflicts, then we
```

Reply:
Apropos of nothing, I note that the former head of the graduate math program at UT streadfastly maintained that the correct plural of "octopus" ought to be "octopodes"